### PR TITLE
DOC: Add documentation for community contributions

### LINKED ISSUE #47: Add `CONTRIBUTING.md` documentation
Document contributing guidelines, including:

- High-level overview of components of each workflow;
- Minimal steps for testing with ITK external module

### PR DESCRIPTION
**Pull request by tbirdso**
Closes #47

### CODE REVIEW

File: docs/CONTRIBUTING.md
Comment (jhlegarreta) [inline docs/CONTRIBUTING.md:26]: "A computer" instead of "PC"?

File: docs/CONTRIBUTING.md
Comment (jhlegarreta) [inline docs/CONTRIBUTING.md:29]: Same as before concerning "PC"

File: docs/CONTRIBUTING.md
Comment (jhlegarreta) [inline docs/CONTRIBUTING.md:30]: Apple uses the case "macOS"

File: docs/CONTRIBUTING.md
Comment (jhlegarreta) [inline docs/CONTRIBUTING.md:32]: Python 3.7 EOL is expected within 6 months. We may start thinking of deprecating it.

File: docs/CONTRIBUTING.md
Comment (jhlegarreta) [inline docs/CONTRIBUTING.md:67]: Github -> GitHub

File: docs/CONTRIBUTING.md
Comment (jhlegarreta) [inline docs/CONTRIBUTING.md:74]: We can use 1. for all numbers to avoid requiring to change them if ever a new point is raised. Markdown takes care of counting.

File: docs/CONTRIBUTING.md
Comment (jhlegarreta) [inline docs/CONTRIBUTING.md:84]: Same as before concerning the numbering.

File: docs/CONTRIBUTING.md
Comment (jhlegarreta) [inline docs/CONTRIBUTING.md:89]: If a change may be integrated into an existing workflow so that other ITK external modules can benefit from beyond the provided functionality, then changes to existing workflows must be privileged. Similarly, workflows (...)"

File: docs/CONTRIBUTING.md
Comment (jhlegarreta) [inline docs/CONTRIBUTING.md:91]: resuable -> reusable

File: docs/CONTRIBUTING.md
Comment (jhlegarreta) [inline docs/CONTRIBUTING.md:178]: Same a before concerning the numbering.

Comment (jhlegarreta) [inline docs/CONTRIBUTING.md:178]: Maybe link to https://github.com/InsightSoftwareConsortium/ITK/blob/23da3dfa79152648f3d38050ee147df47f41ce0a/CONTRIBUTING.md#commit-messages as guidance on what the ITK ecosystem commit messages should contain.

File: docs/CONTRIBUTING.md
Comment (jhlegarreta) [inline docs/CONTRIBUTING.md:180]: You have used underscores for italics previously (_manage continuous integration_). Can we use the same symbol consistently?

File: docs/CONTRIBUTING.md
Comment (jhlegarreta) [inline docs/CONTRIBUTING.md:161]: Same as before concerning the numbering.

File: docs/CONTRIBUTING.md
Comment (jhlegarreta) [inline docs/CONTRIBUTING.md:164]: Despite not being essential, mind the case: https://github.com/InsightSoftwareConsortium/ITKBSplineGradient

File: docs/CONTRIBUTING.md
Comment (jhlegarreta) [inline docs/CONTRIBUTING.md:103]: IMO, in reality, the burden is on the contributor, as ideally the change should be demonstrated in a remote of their choice/for which they need the change, and have forked and where they will trigger the CI via another PR that updates the commit hash of the reusable action. Indeed, you use this _workflow_ :smile: below.

File: docs/CONTRIBUTING.md
Comment (jhlegarreta) [inline docs/CONTRIBUTING.md:99]: May the "At the time of writing (...)" be split somehow? Looks long/difficult to follow.

Also, is "At the time of writing" complete? i.e. Did you mean "At the time of writing the workflow"? "At the time of writing this (guide)"/"At this time"?

File: docs/CONTRIBUTING.md
Comment (jhlegarreta) [inline docs/CONTRIBUTING.md:104]: "from" instead of "of"?

File: docs/CONTRIBUTING.md
Comment (jhlegarreta) [inline docs/CONTRIBUTING.md:113]: An unimportant style remark: why is the line breaking inconsistent?

File: docs/CONTRIBUTING.md
Comment (jhlegarreta) [inline docs/CONTRIBUTING.md:114]: `ITKSplitComponents` to match the style you have used to refer to this repository?

File: docs/CONTRIBUTING.md
Comment (jhlegarreta) [inline docs/CONTRIBUTING.md:116]: Same as above.

File: docs/CONTRIBUTING.md
Comment (jhlegarreta) [inline docs/CONTRIBUTING.md:145]: Should this be the name of the branch or the commit hash?

File: docs/CONTRIBUTING.md
Comment (jhlegarreta) [inline docs/CONTRIBUTING.md:150]: Same as above.

File: docs/CONTRIBUTING.md
Comment (jhlegarreta) [inline docs/CONTRIBUTING.md:161]: First time the GHA acronym is used. Should be defined early (and maybe reused to save typing),

File: docs/CONTRIBUTING.md
Comment (jhlegarreta) [inline docs/CONTRIBUTING.md:120]: Not sure if ">" is preferred over "$".

**Review by jhlegarreta — COMMENTED** [review]
Thanks for doing this Tom.

Can you please add a self-contained description of the contribution in the commit message, body? It helps a lot instead of just saying "Closes #N".

A few in-line suggestions/comments.

File: docs/CONTRIBUTING.md
Comment (tbirdso) [reply docs/CONTRIBUTING.md:92]: Could you please clarify what you mean by this suggestion?

File: docs/CONTRIBUTING.md
Comment (tbirdso) [reply docs/CONTRIBUTING.md:170]: I'm not sure I understand the suggestion?

File: docs/CONTRIBUTING.md
Comment (tbirdso) [reply docs/CONTRIBUTING.md]: I have removed "and maintainers" to reflect that the burden is indeed on the contributor.

File: docs/CONTRIBUTING.md
Comment (tbirdso) [reply docs/CONTRIBUTING.md]: Revised to "As of January 2023" for clarity

File: docs/CONTRIBUTING.md
Comment (tbirdso) [reply docs/CONTRIBUTING.md]: "Required of" is correct

https://idioms.thefreedictionary.com/required+of

File: docs/CONTRIBUTING.md
Comment (tbirdso) [reply docs/CONTRIBUTING.md]: Fixed for now

File: docs/CONTRIBUTING.md
Comment (tbirdso) [reply docs/CONTRIBUTING.md:151]: Can be either the branch name or commit hash, updated docs to reflect this.

File: docs/CONTRIBUTING.md
Comment (jhlegarreta) [reply docs/CONTRIBUTING.md:170]: The font case:
https://github.com/InsightSoftwareConsortium/ITKBSplineGradient
vs
https://github.com/insightSoftwareConsortium/ITKbsplinegradient

File: docs/CONTRIBUTING.md
Comment (jhlegarreta) [reply docs/CONTRIBUTING.md:92]: I meant that it might be more clear to start as:

> If a change may be integrated into an existing workflow so that other ITK external modules can benefit from beyond the provided functionality, then changes to existing workflows must be privileged. Similarly, workflows will need to (...)

File: docs/CONTRIBUTING.md
Comment (tbirdso) [reply docs/CONTRIBUTING.md]: Given that the term "GitHub Actions" is not excessively used in this document I have opted to remove the abbreviation and use the full term throughout for clarity.

File: docs/CONTRIBUTING.md
Comment (tbirdso) [reply docs/CONTRIBUTING.md]: I have replaced with "$" for consistency with ITK's `CONTRIBUTING.md`.

File: docs/CONTRIBUTING.md
Comment (tbirdso) [reply docs/CONTRIBUTING.md]: From testing this it seems that automatic numbering only works if the lines are directly adjacent, updating these lines with script tags in between results in repeated "1." entries. Will revert this section only.

**Review by dzenanz — APPROVED** [review]
Looks good on a glance.

File: docs/CONTRIBUTING.md
Comment (tbirdso) [reply docs/CONTRIBUTING.md:92]: Thanks. If I'm understanding correctly, in my opinion discussions around privileges are already covered in 1) workflow motivation ("Does my workflow provide functionality that will generalize" and 2) PR  setup ("After a repository maintainer reviews, approves, and merges...". The "modifying a workflow" section by itself should focus on the editing process.

Comment (tbirdso) [reply docs/CONTRIBUTING.md]: Ah, I understand now, thanks for clarifying. I agree with your assessment and see where discussion around motives was missing from the "Modifying" section. I've added a couple of notes on what types of modifications will be accepted, please take a look and let me know what you think.

In particular I agree that we want modifications to be "privileged" in the sense that 1) maintainer approval is required and 2) modifications must meet certain criteria in support of the greater ITK external module ecosystem. We want to discourage individual contributors from adding targeted changes that will only benefit their own external modules.

**Comment by dzenanz** [comment]
@tbirdso if a remote module requires some non-standard build dependency (e.g. presence of Perl, or building of specialized library X), it still cannot use this action? It needs to use the old-style code, like in https://github.com/InsightSoftwareConsortium/ITKIOOMEZarrNGFF/blob/5fd54c1099eb44d5d5fb29eb349003b4cb651255/.github/workflows/build-test-package.yml#L39-L40? Is this correct? I really wish they allowed inclusion of yaml files, as per https://github.com/actions/starter-workflows/issues/245.

File: docs/CONTRIBUTING.md
Comment (jhlegarreta) [inline docs/CONTRIBUTING.md:96]: Can use 1. here as well: Markdown will take care of numbering correctly.

File: docs/CONTRIBUTING.md
Comment (jhlegarreta) [reply docs/CONTRIBUTING.md]: Went through the new wording: looks reasonable. Thanks Tom. Resolving.

File: docs/CONTRIBUTING.md
Comment (tbirdso) [reply docs/CONTRIBUTING.md]: Included in squash update 👍

**Review by jhlegarreta — APPROVED** [review]

**Comment by jhlegarreta** [comment]
https://github.com/InsightSoftwareConsortium/ITKRemoteModuleBuildTestPackageAction/pull/51#issuecomment-1376038053 If a "sufficient" number of modules need a given library L, I guess a boolean flag could be added to the workflow. It is somehow suggested in the text Tom proposes that such a feature could be accepted. Not sure how many would be "sufficient". And if they require library L but versions V1 vs V2, we would hit a difficulty there.

**Pull request merged**
